ifacemonitor/update_filter.go

```diff
@@ -0,0 +1,202 @@
+}
+
+type UpdateFilter struct {
+	Time timeshim.Time
+}
+
+type UpdateFilterOp func(filter *UpdateFilter)
+
+func WithTimeShim(t timeshim.Time) UpdateFilterOp {
+	return func(filter *UpdateFilter) {
+		filter.Time = t
+	}
+}
+
+// FilterUpdates filters out updates that occur when IPs are quickly removed and re-added.
+// Some DHCP clients flap the IP during an IP renewal, for example.
+//
+// Algorithm:
+// * Maintain a queue of link and address updates per interface.
+// * When we see a potential flap (i.e. an IP deletion), defer processing the queue for a while.
+// * If the flap resolves itself (i.e. the IP is added back), suppress the IP deletion.
+func (u *UpdateFilter) FilterUpdates(ctx context.Context,
+	addrOutC chan<- netlink.AddrUpdate, addrInC <-chan netlink.AddrUpdate,
+	linkOutC chan<- netlink.LinkUpdate, linkInC <-chan netlink.LinkUpdate) {
+
+	logrus.Debug("FilterUpdates: starting")
+	var timerC <-chan time.Time
+
+	type timestampedUpd struct {
+		ReadyAt time.Time
+		Update  interface{} // AddrUpdate or LinkUpdate
+	}
+
+	updatesByIfaceIdx := map[int][]timestampedUpd{}
+
+	for {
+		select {
+		case <-ctx.Done():
+			logrus.Info("FilterUpdates: Context expired, stopping")
+			return
+		case linkUpd := <-linkInC:
+			idx := linkUpd.Index
+			if len(updatesByIfaceIdx[int(idx)]) == 0 {
+				// We do see link updates as part of a flap but we've only seen them after the IP is
+				// flapped down.  Avoid delaying unrelated link updates.
+				logrus.Debug("FilterUpdates: link change with empty queue, short circuit.")
+				linkOutC <- linkUpd
+				continue
+			}
+			// Otherwise, we might have a flap in progress, delay the update.
+			updatesByIfaceIdx[int(idx)] = append(updatesByIfaceIdx[int(idx)],
+				timestampedUpd{
+					ReadyAt: u.Time.Now().Add(FlapDampingDelay),
+					Update:  linkUpd,
+				})
+		case addrUpd := <-addrInC:
+			idx := addrUpd.LinkIndex
+			oldUpds := updatesByIfaceIdx[idx]
+
+			var readyToSendTime time.Time
+			if addrUpd.NewAddr {
+				logrus.WithField("addr", addrUpd.LinkAddress).Debug("FilterUpdates: got address ADD")
+				if len(oldUpds) == 0 {
+					// This is an add for a new IP and there's nothing else in the queue for this interface.
+					// Short circuit.  We care about flaps where IPs are temporarily removed so no need to
+					// delay an add.
+					logrus.Debug("FilterUpdates: add with empty queue, short circuit.")
+					addrOutC <- addrUpd
+					continue
+				}
+
+				// Else, there's something else in the queue, need to process the queue...
+				logrus.Debug("FilterUpdates: add with non-empty queue.")
+				// We don't actually need to delay the add itself so we don't set any delay here.  It will
+				// still be queued up behind other updates.
+				readyToSendTime = u.Time.Now()
+			} else {
+				// Got a delete, it might be a flap so queue the update.
+				logrus.WithField("addr", addrUpd.LinkAddress).Debug("FilterUpdates: got address DEL")
+				readyToSendTime = u.Time.Now().Add(FlapDampingDelay)
```

Comment:
You might have an ADD in the queue because the ADD previously got blocked behind a different update so I think it makes sense to discard that earlier ADD, which we now know is out of date.